### ISSUE TITLE
Use higher uid/git for Vexim by default

### ISSUE THREAD

**Issue by rimas-kudelis** [vexim/vexim2#80]
In README.md, we currently instruct our users to create a system (low-uid) user for Vexim. This might not be such a good idea after all (e.g. Dovecot disallows logging in for users with uid lower than 500 by default – and for a good reason!). I think we should update the readme to instruct how to create a user with higher uid by by default, e.g. with uid between 500 and 999, or something.

At least in Debian, normal users get uid's starting from 1000. We probably want our user to be a bit special, but not as special as to have a really low uid.

**Comment by Udera** [comment]
Are there other consequences related with the UID (remote login, ...)?

**Comment by rimas-kudelis** [comment]
Don't really know, I can only quote some sources. For example, LSB [specifies](http://refspecs.linuxfoundation.org/LSB_5.0.0/LSB-Core-generic/LSB-Core-generic/uidrange.html) the following: 
> The system User IDs from 0 to 99 should be statically allocated by the system, and shall not be created by applications.
> 
> The system User IDs from 100 to 499 should be reserved for dynamic allocation by system administrators and post install scripts using useradd.

Meanwhile Debian Policy Manual [extends](https://www.debian.org/doc/debian-policy/ch-opersys.html#s9.2.2) the second group up until `uid 1000`. But Dovecot as shipped by Debian still allows us to log in with uid's starting at 500, so I thought why not exploit that gap for our purpose.

Other than that, our user should not be able to log in to the shell anyway (we should use `--disabled-password` and `--disabled-login` when creating it).

**Comment by tflidd** [comment]
CentOS/Redhat give user-ids starting from 500 (https://en.wikipedia.org/wiki/User_identifier). We would be compatible with all systems if we chose a uid between 100-499.

**Comment by Udera** [comment]
CentOS/Redhat give user-ids starting from 500 (https://en.wikipedia.org/wiki/User_identifier). We would be compatible with all systems if we chose a uid between 100-499.

**Comment by Z3po** [comment]
So we can't match all Systems/Services policies...still I think we may stay between 100 and 499.

**Comment by rimas-kudelis** [comment]
The thing is, we don't really need a special uid, it looks better, but there's nothing other than that. I wonder what's worse: having a non-special uid for special use, or (effectively) disabling this protective setting in Dovecot (and potentially other applications).

**Comment by Udera** [comment]
It's more important that the new user can't login via ssh. The actual UID, you could even mention in the documentation to use a 99<UID<500 on RHEL/CentOS.

**Comment by rimas-kudelis** [comment]
The user won't be able to log in via ssh if its password hash field in `/etc/shadow` contains either `*` or `!`, so that is not a problem. At least in debian, `adduser` takes care of this when either `--disabled-password` or `--disabled-login` is used (or both).

**Comment by Udera** [comment]
Perfect.

**Comment by rimas-kudelis** [comment]
What's perfect? I still think we should use an uid of 500 or higher.

**Comment by Udera** [comment]
For me it would be OK to use a gid > 500.

**Comment by xoyteam** [comment]
First I do not understand why it is necessary to use a specific UID?
This setting is specific to each operating system.
I think, because this user is not a real man. It must be allocated a system UID. And what and how it will solve the range of the system.
Enough to make the instructions - how to create a system user, how do you know its UID and how to enter in the configuration.

p.s. Debina Polycy Manual https://www.debian.org/doc/debian-policy/ch-opersys.html#s9.2

> 
> 0-99:
> 
>  Globally allocated by the Debian project, the same on every Debian system. 
> These ids will appear in the passwd and group files of all Debian systems, 
> new ids in this range being added automatically as the base-passwd package is updated.
> 
> Packages which need a single statically allocated uid or gid should use one of these; 
> their maintainers should ask the base-passwd maintainer for ids.
> 
> 100-999:
> 
> Dynamically allocated system users and groups. Packages which need a user or group, 
> but can have this user or group allocated dynamically and differently on each system, 
> should use adduser --system to create the group and/or user. adduser will check for the existence of the user or group, and if necessary choose an unused id based on the ranges specified in adduser.conf.
> 
> 1000-59999:
> 
> Dynamically allocated user accounts. By default adduser will choose UIDs and GIDs 
> for user accounts in this range, though adduser.conf may be used to modify this behavior.

**Comment by rimas-kudelis** [comment]
> First I do not understand why it is necessary to use a specific UID?
> This setting is specific to each operating system.

We do not supply a specific UID or GID. What we do though is tell the users to create a system user (which might fall anywhere between 100 and 999 in Debian, or which we specify to be between 1 and 100 for FreeBSD, I suppose).

> I think, because this user is not a real man. It must be allocated a system UID. And what and how it will solve the range of the system.
> Enough to make the instructions - how to create a system user, how do you know its UID and how to enter in the configuration.

In general, I would agree. But if you have read my rationale above, you know that Dovecot does not allow anyone with UID lower than 500 to log in. Since all virtual users are treated as having the UID we create here, we have to take that into account, hence my suggestion to limit the possible UID's with 500 from below. It doesn't mean we're going to specify a particular UID to use, it just means we're setting some limits.

**Comment by xoyteam** [comment]
> But if you have read my rationale above, you know that Dovecot does not allow anyone with UID lower than 500 to log in.
It also applies in the Debian installation of Dovecot package?
Maybe it was a mistake debian maintainer?

**Comment by rimas-kudelis** [comment]
Yes, it applies to Debian as well. It's an upstream policy, which is configurable, but I don't see a reason why we would need UID's that low. Here's the exact quote:
> By default Dovecot allows users to log in only with UID numbers 500 and above. This check tries to make sure that no-one can ever log in as daemons or other system users. If you're using an UID lower than 500, you'll need to change the `first_valid_uid` setting.

And here's the link: http://wiki2.dovecot.org/UserIds#UIDs

**Comment by rimas-kudelis** [comment]
Come to think of it, Dovecot not disabling logins up until `uid=1000` in Debian is probably a bug and an oversight in a Debian package, not a feature.
I kind of like the fact that we're creating a system user after all, so perhaps instead of changing that, we should just add a note to our Dovecot manual (and perhaps to the README as well)?

**Comment by rimas-kudelis** [comment]
Or we can just update the instruction and let the user create an account in the unpriviledged space. Damn, I'm soooo unsure which is better...

**Comment by Udera** [comment]
Just followed the link you were giving. This is a bit worrying:

> The important thing to consider with your UID allocation policy is that if Dovecot has a security hole in its IMAP or POP3 implementation, the attacker can read mails of other people who are using the same UID. So clearly the most secure way is to allocate a different UID for each user. It can however be a bit of a pain and OSes don't always support more than 65536 UIDs.

**Comment by rimas-kudelis** [comment]
Yes it is worrying, but what other options do we have? Using uid's of non-existing users doesn't look like an attractive solution, creating separate UNIX users for every Vexim user either.

**Comment by Udera** [comment]
We could create a bunch of dummy-users and link the `uid`s to accounts. But that's a very ugly and cumbersome workaround (that I don't want to do).

back to the uid: I think it is good to have a special `uid` outside the normal users range because you handle these users with extra care (e.g. that you didn't forget to give a password to this user). And there are so many systems around that the user has to verify the `first_valid_uid`.
`useradd -r -g vmail vmail` creates a system user that should be working on all Linux systems. This dynamic system user is different on Linux distribution, it would make the howto more complicated (choose any uid between 500 and 999 on debian systems, on CentOS between 100-499, on SuSE... if the uid is already used take a different one, ...)

**Comment by rimas-kudelis** [comment]
@Udera I don't understand what you're suggesting. :)

**Comment by Udera** [comment]
I'm just suggesting to create a system user `vmail` and take this UID.

**Comment by rimas-kudelis** [comment]
I see, so you mean an account with a privileged uid. Okay.
I was kind of leaning towards an unprivileged one, but I do have doubts regarding whether securing system daemon-by-daemon is the right approach. If your user which is not supposed to log in (e.g. syslog) has a password set, then perhaps you have deeper issues than just ability of someone knowing that password to check that user's e-mail.

However, I don't think I found a way to add a privileged system user in FreeBSD easily. That will still require quite an ugly command (#121).

**Comment by Udera** [comment]
http://www.afterschool.dk/documents/freebsd/freebsd-postfix-guide/
````
pw groupadd vmail -g 5001 
pw useradd vmail -u 5001 -g 5001 -s/sbin/nologin -d/dev/null
````
On debian, you could as well use such high `uid`s that won't interfere with normal users.

**Comment by rimas-kudelis** [comment]
That's unprivileged, and there's no need to specify the uid nor gid if we're going down that road. I was talking about low id's.

**Comment by xoyteam** [comment]
> I was kind of leaning towards an unprivileged one, but I do have doubts regarding whether securing system daemon-by-daemon is the right approach. If your user which is not supposed to log in (e.g. syslog) has a password set...

where vexim2 uses the password for the system user?!

> However, I don't think I found a way to add a privileged system user in FreeBSD easily. That will still require quite an ugly command

It is not a problem. It is possible to write for the instruction a one-line script on search of first_valid_uid.
Here the main thing to define a policy.

Forgive me, but I consider that it isn't good to use the fixed uid.

**Comment by rimas-kudelis** [comment]
>>    I was kind of leaning towards an unprivileged one, but I do have doubts regarding whether securing system daemon-by-daemon is the right approach. If your user which is not supposed to log in (e.g. syslog) has a password set...
>
> where vexim2 uses the password for the system user?!

It doesn't.

>>    However, I don't think I found a way to add a privileged system user in FreeBSD easily. That will still require quite an ugly command
>
> It is not a problem. It is possible to write for the instruction a one-line script on search of first_valid_uid.
> Here the main thing to define a policy.

Yes, see #121.

> Forgive me, but I consider that it isn't good to use the fixed uid.

We aren't using a fixed uid. But we need to know the uid and gid once the user is created, because we have to put it into the SQL setup script. Check out our README.md and you'll see what I mean.

**Comment by xoyteam** [comment]
The talc problem in writing of a script means?
And that I created a bug and I read discussion, I don't understand in what difficulty now? I can be engaged in correction #121.

**Comment by rimas-kudelis** [comment]
Like I said, we just have to decide: do we create a user with low/unprivileged uid (<1000 in most cases), or not. Depending on that, the commands slightly differ. Adding a user with unprivileged uid is easy in all systems, but adding one with a lower uid is different: `adduser` and `useradd` in Debian have a `--system` switch, but neither `pw useradd`, nor `adduser` in FreeBSD have one.

**Comment by xoyteam** [comment]
And once again I will repeat that this basic decision, and remaining - difficulties of implementation.

On how many I understand, consider all arguments privileged UID.
Well, I will try to realize an example for FreeBSD.

**Comment by Udera** [comment]
I'm currently going through the documentation to set up a fresh debian 8. Can't we just choose a UID/GID? This way, we could save a few steps by setting the uid/gid manually in the database and config file. Normally the mail server is one of the first things to be installed and if the UID collides, the user can still use a different one. But for the majority of users, this will simplify the installation process.

Anything we fancy? The last possible one `999` or the first catchy-one `555`?

**Comment by rimas-kudelis** [comment]
I don't think it's a good idea really. Perhaps we should leave the uid as-is and retarget this issue to 3.0 milestone instead?

**Comment by Udera** [comment]
Let's retarget this. It's nothing critical that must be fixed right now.
